Docs for new commondata format (#1708)
NNPDF/nnpdf: doc/sphinx/source/data/new-commondata.rst

> Organization following the naming convention

Reviewer: Is the convention enforced mechanically? Is it needed to process the names?

Author: > Is it needed to process the names?

Yes. The folder in which the data is will always be `name.rsplit("_", 1)[0]` and the choice of the observable `name.rsplit("_", 1)[1]

--------------------------------------------

All dataset in the new data format follow the exact same naming convention::

    <experiment>_<process>_<energy>{_<extras>}_<observable>

> The data is contained in folders, each containing one single hepdata publication. 

Reviewer: I assume we can do other things that are not hepdata?

Author: Of course vp doesn't know where the data come from but I think the idea is that you open a folder and it comes from one single hepdata publication.

In the odd case in which it doesn't we hope the person implement it add comments or something explaining what happened there.

Reviewer: The comment was more in the direction that these docs should be rather general.

In all cases one can reconstruct the name of the folder by separating the observable name on the last ``_``, i.e., the folder will always be named::

    <experiment>_<process>_<energy>{_<extras>}

Where all observables contained in one hepdata entry are separated by their observable name.

> Metadata Format

Reviewer: At some point before here we should discuss what is required (metadata, kinematics) and what is optional (everything else).

Author: I don't agree with the everything else. I think hepdata is required and then if it is empty one can add "NA" or whatever. Same for version. Or maybe you are referring to that with `metadata`.

The reason is that I would like to avoid a situation in which a dataset is implemented (or amended) with that information missing even if the cost is a slightly more cumbersome implementation for say a positivity dataset.

Organization following the naming convention
--------------------------------------------

All dataset in the new data format follow the exact same naming convention::

    <experiment>_<process>_<energy>{_<extras>}_<observable>

The data is contained in folders, each containing one single hepdata publication. 
In all cases one can reconstruct the name of the folder by separating the observable name on the last ``_``, i.e., the folder will always be named::

    <experiment>_<process>_<energy>{_<extras>}

Where all observables contained in one hepdata entry are separated by their observable name.

Metadata Format
---------------

> The new data format is center around the file ``metadata.yaml`` contained inside the set folder.

Reviewer 2: ```suggestion
The new data format is centered around the file ``metadata.yaml`` contained inside the set folder.
```

> This ``metadata.yaml`` file contains first portion of general information which might be shared by several sets and a list of ``implemented_observables`` which define the separate observables.

Reviewer 2: ```suggestion
This ``metadata.yaml`` file contains a first portion of general information which might be shared by several sets and a list of ``implemented_observables`` which define the separate observables.
```

..  code-block:: yaml

    setname: "EXPERIMENT_PROCESS_ENERGY

    version: 1
    version_comment: "Initial implementation"

    # References
    arXiv:
        url: ""
    iNSPIRE:
        url: "https://inspirehep.net/literature/302822"
    hepdata:
        url: "https://www.hepdata.net/record/ins302822"
        version: 1

    nnpdf_metadata:
        nnpdf31_process: "PROCESS"
        experiment: "EXPERIMENT_NAME"

    implemented_observables:
    - observable_name: "OBS"
        observable:
            description: "Description of the observable"
            label: "Latex label for the observable"
            units: "[u]"
        ndata: n_of_datapoints

>         tables: (optional) [n, j, k] # corresponding tables in the hepdata entry

Reviewer 2: ```suggestion
        tables: [n, j, k] # (optional) corresponding tables in the hepdata entry
```
similar below

        npoints: (optional) [n, j, k] # number of points per table

        # Plotting information
        dataset_label: "Label to be used in reports"
        plot_x: variable_1
        figure_by:
            - variable_2
        kinematic_coverage:
            variable_1: var_1
            variable_2: var_2

        kinematics:
            variables:

>                 var_1: {description: "Description of var", label: latex, units: "u"}

Reviewer 2: ```suggestion
                var_1: {description: "Description of var", label: "latex", units: "u"}
```
to be consistent with the above - also do I need the `[` inside units?

Reviewer 2: @scarlehoff  how about the units?

                    - uncertainties_3.yaml

        # The theory field is always optional
        theory: 
            FK_tables:
                - - DYE605
            operation: 'null'

Versioning
~~~~~~~~~~

The initial version of a dataset should be set to ``version: 1``.
Any change on a dataset should be *always* accompanied of a version bump and a ``version_comment`` explaining the update.
This will allow to keep an exact tracking of all changes to every dataset even if they change over time.

Variants
~~~~~~~~

In some ocassions we might want to maintain two variations of the same observable.
For instance, we might have two incompatible sources of uncertainties. In such case a variant can be added.
The syntax of the ``variants`` is:

Theory
~~~~~~

The theory field defines how predictions for the dataset are to be computed.
It includes two entries:

- ``FK_tables``: this is a list of lists which defines the fktables to be loaded. The outermost list are the operands (in case an operation is needed to recover the observable, more on that below). The innermost list are the grids that are to be concatenated in order to form the operands.

> - ``operaton``: operation to be applied in order to compute the observable

Reviewer 2: we definitely need some cross link here ... I don't know if and where the operations are explained, though

Example:

..  code-block:: yaml
            theory: 
            FK_tables:
                - - Z_contribution
                  - Wp_contribution
                  - Wm_total
                - - total_xs
            operation: 'ratio'

In this case the ``fktables`` for the Z, W+ and W- contributions will be concatenated (the dataset might include predictions for all three contributions).
After that, the final observable will be computed by taking the ratio of the concatenation of all those observables and the total cross section (``total_xs``).

> ..  code-block:: yaml

Reviewer 2: this paragraph needs to move back above the theory section

    data_uncertainties:
        - uncertainties.yaml

    variants:
        name_of_the_variant:
            data_uncertainties:
                - uncertainties.yaml
                - extra_uncertainties.yaml
        another_variant:
            data_uncertainties:
                - different_uncertainties.yaml

When loading this dataset with no variant only the ``uncertainties.yaml`` file will be read.
Instead, when choosing ``variant: name_of_the_variant``, both ``uncertainties.yaml`` and  ``extra_uncertainties.yaml`` will be loaded.
Note that if we want to substitute the default set of uncertainties we just need to not include it in the variant (as done in ``another_variant``).

Data
----

> The format of the data is a ``yaml`` file with an entry ```data_central``` which is a list for all values for all bins.

Reviewer 2: I guess we should stress somehow more that this is the crucial part

..  code-block:: yaml

    data_central:
        - val1
        - val2
        - val3

> Uncertainties:
> --------------

Reviewer 2: ```suggestion
Uncertainties
-------------
```

Note that in the ``metadata.yaml`` the ``data_uncertainties`` entry is given as a list. 
When using more than one uncertainty file they will be concatenated. 
This allows the user the flexibility of creating variants where only a subset of the uncertainties are modified.

The format of the uncertainty files is of two fields, a ``definitions`` field that contains metadata about all the uncertainties (their name, their treatment (``ADD`` or ``MULT``) and their type) and a second field ``bins`` which is a list of mappings with as many entries as the `data_central` with the named uncertainties.

..  code-block:: yaml

    definitions:
        stat:
            description:
            treatment:
            type:
        error_name:
            description:
            treatment:
            type:
        error_name_2:
            description:
            treatment:
            type:
    bins:
        - stat:
        error_name:
        error_name_2:

Kinematics:
-----------
The kinematics file follow a convention very similar to the uncertainties file, where the ``definitions`` field is skipped since that information is already contained in the parent ``metadata.yaml`` file.

> Therefore, we have a list of ``bins`` (of the same size as the previous lists) and for each entry we have the information of all the variables.

Reviewer 2: ```suggestion
Therefore, we have a list of ``bins`` (of the same size as the list for `data_central`) and for each entry we have the information of all the variables.
```

..  code-block:: yaml

    bins:

>         - var:
>             min: 0
>             max: 1
>             mid: 0.5
>         var_2:

Reviewer 2: ```suggestion
        - var:
            min: 0
            max: 1
            mid: 0.5
          var_2:
```
same as above

When a ``legacy`` variant is available, the usage of the old name automatically enables such variants.

Metadata Format
---------------

This ``metadata.yaml`` file contains a first portion of general information which might be shared by several sets and a list of ``implemented_observables`` which define the separate observables.

..  code-block:: yaml

    setname: "EXPERIMENT_PROCESS_ENERGY"

    version: 1
    version_comment: "Initial implementation"

    # References
    arXiv:
        url: ""
    iNSPIRE:
        url: "https://inspirehep.net/literature/302822"
    hepdata:
        url: "https://www.hepdata.net/record/ins302822"
        version: 1

    nnpdf_metadata:
        nnpdf31_process: "PROCESS"
        experiment: "EXPERIMENT_NAME"

    implemented_observables:
    - observable_name: "OBS"
        observable:
            description: "Description of the observable"
            label: "Latex label for the observable"
            units: "[u]"
        ndata: n_of_datapoints
        tables: [n, j, k] # (optional) corresponding tables in the hepdata entry
        npoints: [n, j, k] # (optional) number of points per table

> 

Author: ```suggestion
        process_type: INC # for instance, INC, JET, DIJET, etc
```

..  code-block:: yaml

    setname: "EXPERIMENT_PROCESS_ENERGY"

    version: 1
    version_comment: "Initial implementation"

    # References
    arXiv:
        url: ""
    iNSPIRE:
        url: "https://inspirehep.net/literature/302822"
    hepdata:
        url: "https://www.hepdata.net/record/ins302822"
        version: 1

    nnpdf_metadata:
        nnpdf31_process: "PROCESS"
        experiment: "EXPERIMENT_NAME"

    implemented_observables:
    - observable_name: "OBS"
        observable:
            description: "Description of the observable"
            label: "Latex label for the observable"
            units: "[u]"
        ndata: n_of_datapoints
        tables: [n, j, k] # (optional) corresponding tables in the hepdata entry
        npoints: [n, j, k] # (optional) number of points per table

        # Plotting information
        plotting:
            dataset_label: "Label to be used in reports"
            plot_x: variable_1
            figure_by:
                - variable_2

>             kinematic_coverage:
>                 variable_1: var_1
>                 variable_2: var_2
>                 variable_3: var_3

Author: ```suggestion
        kinematic_coverage:
            variable_1: var_1
            variable_2: var_2
            variable_3: var_3
```

        name_of_the_variant:
            data_uncertainties:
                - uncertainties.yaml
                - extra_uncertainties.yaml
        another_variant:
            data_uncertainties:
                - different_uncertainties.yaml

When loading this dataset with no variant only the ``uncertainties.yaml`` file will be read.
Instead, when choosing ``variant: name_of_the_variant``, both ``uncertainties.yaml`` and  ``extra_uncertainties.yaml`` will be loaded.
Note that if we want to substitute the default set of uncertainties we just need to not include it in the variant (as done in ``another_variant``).

Data
----

The format of the data is a ``yaml`` file with an entry ```data_central``` which is a list for all values for all bins.

..  code-block:: yaml

    data_central:
        - val1
        - val2
        - val3

Uncertainties
-------------

The uncertainties are (also) ``.yaml`` files. 
Note that in the ``metadata.yaml`` the ``data_uncertainties`` entry is given as a list. 
When using more than one uncertainty file they will be concatenated. 
This allows the user the flexibility of creating variants where only a subset of the uncertainties are modified.

> The format of the uncertainty files is of two fields, a ``definitions`` field that contains metadata about all the uncertainties (their name, their treatment (``ADD`` or ``MULT``) and their type) and a second field ``bins`` which is a list of mappings with as many entries as the `data_central` with the named uncertainties.

Author: ```suggestion
The format of the uncertainty files is of two fields, a ``definitions`` field that contains metadata about all the uncertainties (their name, their treatment (``ADD`` or ``MULT``) and their type) and a second field ``bins`` which is a list of mappings with as many entries as the `data_central` with the named uncertainties.

Note that, regardless of their treatment type, the uncertainties should always be written as absolute values and not relative to the data values.
```

The format of the uncertainty files is of two fields, a ``definitions`` field that contains metadata about all the uncertainties (their name, their treatment (``ADD`` or ``MULT``) and their type) and a second field ``bins`` which is a list of mappings with as many entries as the `data_central` with the named uncertainties.

> Note that, regardless of their treatment type, the uncertainties should always be written as absolute values and not relative to the data values.

Reviewer 3: This statement, I think, is not fully correct.
